apps/backend/src/app/modules/session/run/run-processor.class.ts

```diff
@@ -154,159 +191,176 @@ export class RunProcessor {
+    //
+    // Main track or bonus track runs always have a dedicated end zone.
+    // Remember, in all cases the /end request sent on hitting the end zone
+    // doesn't generate a timestamp.
+    if (!checkpointsRequired) return;
+
+    if (this.session.trackType === TrackType.STAGE) {
+      const isLastSegment =
+        this.session.trackNum === this.zones.tracks.main.zones.segments.length;
+
+      if (this.zones.tracks.main.stagesEndAtStageStarts || isLastSegment) {
+        if (timestamps.length !== checkpoints.length) {
+          throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
+        }
+      } else {
+        if (timestamps.length !== checkpoints.length - 1) {
+          throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
+        }
+      }
+    } else {
+      if (timestamps.length !== checkpoints.length) {
+        throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
+      }
+    }
+  }
+
+  /** @throws {RunValidationError} */
+  validateReplayHeader() {
+    const { session, replayHeader: header } = this;
+
+    if (header.trackType !== session.trackType)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (header.trackNum !== session.trackNum)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (header.magic !== ReplayFile.REPLAY_MAGIC)
+      throw new RunValidationError(ErrorType.BAD_META);
 
-    // If stagesEndAtStageStarts is true then then timestamps should
-    // contain every checkpoint, since the end zone is either the first cp
-    // of the next segment, or the main track's end zone.
-    // If false, the end zone is the final checkpoint of the current segment,
-    // so the timestamps should contain every checkpoint except the last.
-    // Remember, the /end request sent on hitting the end zone doesn't have a
-    // timestamp.
     if (
-      this.trackType === TrackType.STAGE &&
-      !this.zones.tracks.main.stagesEndAtStageStarts
+      header.mapHash.toUpperCase() !==
+      session.mmap.currentVersion.bspHash.toUpperCase()
     )
-      expectedTimestamps -= 1;
+      throw new RunValidationError(ErrorType.BAD_META);
 
-    if (checkpointsRequired && timestamps.length !== expectedTimestamps)
-      throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
-  }
+    if (header.mapName !== session.mmap.name)
+      throw new RunValidationError(ErrorType.BAD_META);
 
-  processReplayFileHeader() {
-    try {
-      this.replay = {
-        ...this.replayFile.readHeader(),
-        overallStats: null,
-        zoneStats: [],
-        frames: []
-      };
-    } catch {
-      throw new RunValidationError(ErrorType.BAD_REPLAY_FILE);
+    if (header.playerSteamID !== this.user.steamID)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (header.gamemode !== session.gamemode)
+      throw new RunValidationError(ErrorType.BAD_META);
+
+    if (!approxEq(header.tickInterval, Tickrates.get(session.gamemode)))
```

Comment:
Tick interval should be exactly the official value I would think

Reply:
If this is a matter of float32 vs JavaScript's native doubles, `Tickrates` should be defined as float32.

On a side note I just realized it's called `Tickrates` but stores tick *intervals* (the inverse of tickrate). I think we have a bad habit of using the terms interchangeably in game code too.

Reply:
If this is a matter of float32 vs JavaScript's native doubles, `Tickrates` should be defined as float32.

On a side note I just realized it's called `Tickrates` but stores tick *intervals* (the inverse of tickrate). I think we have a bad habit of using the terms interchangeably in game code too.

Reply:
Didn't know there was a nice way to do this with native JS until now, but turns out [`Math.fround`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/fround) is a thing and does what we're after.

My bad wrt the tickrate/intervals thing, definitely something we should be using precisely, have updated the naming.